### PR TITLE
Force SeaBIOS instead of OVMF-based firmware & some firmware lookup logic changes

### PR DESCRIPTION
**Pull request by mihalicyn**
- We want our users to step out from EDK2-based CSM firmwares to SeaBIOS as these firmwares were removed from upstream (see https://bugzilla.tianocore.org/show_bug.cgi?id=4588).

- @tomponline suggested to extend firmware lookup algorithm to handle multiple paths specified in the `LXD_OVMF_PATH`.

- Do some renaming

- Forbid `security.csm` for arches except `x86_64`

### CODE REVIEW

**Comment by mihalicyn** [comment]
I have tested this with a VM which was created on the LXD 5.19 and then upgraded to the LXD version with these patches. It works without any problems.

File: doc/environment.md [hunk 1]
```diff
@@ -36,6 +36,7 @@ Name                            | Description
 `LXD_LXC_TEMPLATE_CONFIG`       | Path to the LXC template configuration directory
 `LXD_SECURITY_APPARMOR`         | If set to `false`, forces AppArmor off
 `LXD_UNPRIVILEGED_ONLY`         | If set to `true`, enforces that only unprivileged containers can be created. Note that any privileged containers that have been created before setting LXD_UNPRIVILEGED_ONLY will continue to be privileged. To use this option effectively it should be set when the LXD daemon is first set up.
-`LXD_OVMF_PATH`                 | Path to an OVMF build including `OVMF_CODE.fd` and `OVMF_VARS.ms.fd`
+`LXD_OVMF_PATH`                 | Path to an OVMF build including `OVMF_CODE.fd` and `OVMF_VARS.ms.fd` (to be deprecated, please use `LXD_QEMU_FW_PATH` instead)
```
Comment (simondeziel) [inline doc/environment.md:39]:
```suggestion
`LXD_OVMF_PATH`                 | Path to an OVMF build including `OVMF_CODE.fd` and `OVMF_VARS.ms.fd` (deprecated, please use `LXD_QEMU_FW_PATH` instead)
```
This feels a little stronger.

File: doc/environment.md [hunk 2]
```diff
@@ -36,6 +36,7 @@ Name                            | Description
...
+`LXD_QEMU_FW_PATH`              | Path (or list of paths) to a firmwares (OVMF, SeaBIOS) to be used by Qemu
```
Comment (simondeziel) [inline doc/environment.md:40]:
```suggestion
`LXD_QEMU_FW_PATH`              | Path (or `:` separated list of paths) to firmwares (OVMF, SeaBIOS) to be used by Qemu
```

**Review by simondeziel — APPROVED** [review]
LGTM with only tiny suggestions.

File: doc/environment.md (hunk 1 above)
Comment (mihalicyn) [reply doc/environment.md]: Thanks, fixed ;-)

File: doc/environment.md (hunk 2 above)
Comment (mihalicyn) [reply doc/environment.md]: Thanks, fixed!

File: doc/environment.md [hunk 3]
```diff
@@ -36,6 +36,7 @@ Name                            | Description
 `LXD_LXC_TEMPLATE_CONFIG`       | Path to the LXC template configuration directory
 `LXD_SECURITY_APPARMOR`         | If set to `false`, forces AppArmor off
 `LXD_UNPRIVILEGED_ONLY`         | If set to `true`, enforces that only unprivileged containers can be created. Note that any privileged containers that have been created before setting LXD_UNPRIVILEGED_ONLY will continue to be privileged. To use this option effectively it should be set when the LXD daemon is first set up.
-`LXD_OVMF_PATH`                 | Path to an OVMF build including `OVMF_CODE.fd` and `OVMF_VARS.ms.fd`
+`LXD_OVMF_PATH`                 | Path to an OVMF build including `OVMF_CODE.fd` and `OVMF_VARS.ms.fd` (deprecated, please use `LXD_QEMU_FW_PATH` instead)
+`LXD_QEMU_FW_PATH`              | Path (or `:` separated list of paths) to firmwares (OVMF, SeaBIOS) to be used by Qemu
```
Comment (simondeziel) [inline doc/environment.md:40]:
```suggestion
`LXD_QEMU_FW_PATH`              | Path (or `:` separated list of paths) to firmwares (OVMF, SeaBIOS) to be used by QEMU
```

Sorry for not catching this on the first review :/
The all in caps version seems to be what upstream uses and also what we have in our docs.

**Review by simondeziel — APPROVED** [review]

File: doc/environment.md (hunk 3 above)
Comment (mihalicyn) [reply doc/environment.md]: it's my bad. Thanks, fixed!

File: lxd/instance/drivers/driver_qemu.go [hunk 4]
```diff
@@ -784,12 +785,27 @@ func (d *qemu) Rebuild(img *api.Image, op *operations.Operation) error {
...
+func (*qemu) fwPath(filename string) string {
+	qemuFwPaths := "/usr/share/OVMF"
+	if os.Getenv("LXD_QEMU_FW_PATH") != "" {
+		qemuFwPaths = os.Getenv("LXD_QEMU_FW_PATH")
+	} else if os.Getenv("LXD_OVMF_PATH") != "" {
+		qemuFwPaths = os.Getenv("LXD_OVMF_PATH")
 	}
 
-	return "/usr/share/OVMF"
+	for _, path := range strings.Split(qemuFwPaths, ":") {
```
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go:796]: We could use the `qemuFwPaths()` function described above here too.

File: lxd/apparmor/instance.go [hunk 5]
```diff
@@ -182,14 +182,19 @@ func instanceProfile(sysOS *sys.OS, inst instance) (string, error) {
 			return "", err
 		}
 
-		ovmfPath := "/usr/share/OVMF"
-		if os.Getenv("LXD_OVMF_PATH") != "" {
-			ovmfPath = os.Getenv("LXD_OVMF_PATH")
+		qemuFwPaths := "/usr/share/OVMF"
```
[the remark this answers is not in the document]
Comment (tomponline) [reply lxd/apparmor/instance.go:185]: We could eval the symlinks in this function too as both potential call sites do that. The function signature would need to be changed to return an error then.

File: lxd/instance/drivers/driver_qemu.go [hunk 6]
```diff
@@ -3041,43 +3053,51 @@ func (d *qemu) generateQemuConfigFile(cpuInfo *cpuTopology, mountInfo *storagePo
...
-			// force to use a 4MB firmware
-			ovmfCode = firmwares[0].code
+			// force to use a top-priority firmware
+			vmfCode = firmwares[0].code
 		}
 
 		// Use debug version of firmware. (Only works for "default" (4MB, no CSM) firmware flavor)
-		if shared.IsTrue(d.localConfig["boot.debug_edk2"]) && ovmfCode == ovmfGenericFirmwares[0].code {
-			ovmfCode = ovmfDebugFirmware
+		if shared.IsTrue(d.localConfig["boot.debug_edk2"]) && vmfCode == vmGenericFirmwares[0].code {
```
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go:3090]: Do we need to also block the usage of boot.debug_edk2 with security.csm now?

Comment (mihalicyn) [reply lxd/instance/drivers/driver_qemu.go:3097]: yep, fixed!

File: lxd/instance/drivers/driver_qemu.go (hunk 4 above)
Comment (mihalicyn) [reply lxd/instance/drivers/driver_qemu.go]: Have done.

File: lxd/apparmor/instance.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (mihalicyn) [reply lxd/apparmor/instance.go]: Thanks for suggestion! Have done!

**Review by tomponline — APPROVED** [review]
LGTM thanks!

**Comment by simondeziel** [comment]
@mihalicyn I took the liberty to edit the PR description for posterity to reflect the incompatibility with `boot.debug_edk2` and `security.secureboot`.

**Pull request merged**